asf_levies_model/summary.py

```diff
@@ -445,3 +445,193 @@ def process_rebalancing_scenario_bills(
+            "Total cost levied to general taxation": cost_to_tax,
+            "Total policy cost revenue": total_cost,
+        }
+    ]
+
+    df = pd.DataFrame(cost_streams_row)
+
+    return df
+
+
+def set_common_denominators(
+    levies: List,
+    supply_elec: float,
+    supply_gas: float,
+    customers_gas: int,
+    customers_elec: int,
+) -> Dict:
+    """Returns a dictionary of common denominators for each levy.
+
+    Parameters
+    ----------
+    levies : List
+        List of levy objects.
+    supply_elec : float
+        Amount of annual electricity supply to use for rebalancing, in MWh.
+    supply_gas : float
+        Amount of annual gas supply to use for rebalancing, in MWh.
+    customers_gas : int
+        Number of customers paying for gas in a given year to use for rebalancing.
+    customers_elec : int
+        Number of customers paying for electricity in a given year to use for rebalancing.
+
+    """
+    denominator_values = {
+        "supply_elec": supply_elec,
+        "supply_gas": supply_gas,
+        "customers_gas": customers_gas,
+        "customers_elec": customers_elec,
+    }
+
+    return {key: denominator_values for key in [levy.short_name for levy in levies]}
+
+
+def calculate_fuel_poverty_rates(consumers: Dict, scenario_names: List) -> pd.DataFrame:
+    """Returns a DataFrame containing percentage of fuel poor households in each archetype, for a set of rebalancing scenarios.
+
+    Parameters
+    ----------
+    consumers : Dict
+        Dictionary of Consumer objects where key is rebalancing scenario name (str) and value is a List of Consumer objects with tariff attributes corresponding to the rebalancing scenario.
+    scenario_names : List
+        List of rebalancing scenario names (str) being considered.
+
+    Returns
+    -------
+    pd.DataFrame
+        DataFrame where rows are archetypes, columns are rebalancing scenario names and values are percentage that is in fuel poverty.
+    """
+
+    # Consumers summary dataframe
+    tidy_summary = pd.DataFrame()
+    for scenario_name in scenario_names:
+        tidy = pd.concat(
+            [consumer.get_tidy_summary() for consumer in consumers[scenario_name]]
+        )
+        tidy["scenario"] = scenario_name
+        tidy_summary = pd.concat([tidy_summary, tidy])
+
+    # Convert to pivot table
+    fp_gap_pivot = tidy_summary.pivot_table(
+        index=["Name", "scenario"], values="Value", columns="Attribute", aggfunc="first"
+    )
+    fp_gap_pivot = fp_gap_pivot[
+        ["archetype", "net_income_decile", "fuel_poverty_gap", "size"]
+    ].reset_index()
+
+    # Create new column to indicate if in fuel poverty
+    fp_gap_pivot["in_fuel_poverty"] = fp_gap_pivot["fuel_poverty_gap"] != 0
+
+    # Helper function to calculate percentage of households in archetype in fuel poverty
+    def _calculate_percentage_in_fuel_poverty(fp_gap_pivot, archetypes, scenario):
+        percentages = []
```

Comment:
These calculations could be refactored to use views with `.loc[]` rather than chained indexing.